File: experiments/03_synth_data_generation/c_src/batch07_string_ops.c
/* Batch 07: String/char-array operations — model char* as Array UInt8 in Lean */

int str_len(const char *s) {
    int len = 0;
    while (s[len] != '\0') {
        len++;
    }
    return len;
}

int str_equal(const char *a, const char *b) {
    int i = 0;
    while (a[i] != '\0' && b[i] != '\0') {
        if (a[i] != b[i]) return 0;
        i++;
    }
    return a[i] == b[i];
}

int is_palindrome(const char *s, int len) {
    for (int i = 0; i < len / 2; i++) {
        if (s[i] != s[len - 1 - i]) return 0;
    }
    return 1;
}

int count_char(const char *s, char c) {
    int count = 0;
    for (int i = 0; s[i] != '\0'; i++) {
        if (s[i] == c) count++;
    }
    return count;
}

void to_upper(char *s) {
    for (int i = 0; s[i] != '\0'; i++) {
        if (s[i] >= 'a' && s[i] <= 'z') {
            s[i] = s[i] - 'a' + 'A';
        }
    }
}

void to_lower(char *s) {
    for (int i = 0; s[i] != '\0'; i++) {
        if (s[i] >= 'A' && s[i] <= 'Z') {
            s[i] = s[i] - 'A' + 'a';
        }
    }
}

int char_is_digit(char c) {
    return c >= '0' && c <= '9';
}

int char_is_alpha(char c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
}

File: experiments/03_synth_data_generation/py_src/batch07_string_ops.py
"""Batch 07: String operations — pure functions on strings/byte arrays."""

def str_len(s: str) -> int:
    count = 0
    for _ in s:
        count += 1
    return count

def str_equal(a: str, b: str) -> bool:
    if len(a) != len(b):
        return False
    for i in range(len(a)):
        if a[i] != b[i]:
            return False
    return True

def is_palindrome(s: str) -> bool:
    n = len(s)
    for i in range(n // 2):
        if s[i] != s[n - 1 - i]:
            return False
    return True

def count_char(s: str, c: str) -> int:
    count = 0
    for ch in s:
        if ch == c:
            count += 1
    return count

def to_upper(s: str) -> str:
    result = []
    for c in s:
        if 'a' <= c <= 'z':
            result.append(chr(ord(c) - ord('a') + ord('A')))
        else:
            result.append(c)
    return ''.join(result)

def to_lower(s: str) -> str:
    result = []
    for c in s:
        if 'A' <= c <= 'Z':
            result.append(chr(ord(c) - ord('A') + ord('a')))
        else:
            result.append(c)
    return ''.join(result)

def char_is_digit(c: str) -> bool:
    return '0' <= c <= '9'

def char_is_alpha(c: str) -> bool:
    return ('a' <= c <= 'z') or ('A' <= c <= 'Z')
